prog/dftb+/lib_dftbplus/mainapi.F90

```diff
@@ -188,10 +191,433 @@ function nrOfAtoms()
+          call error("DFTB+U is not supported by DFTB+ API")
+       elseif(allocated(onSiteElements)) then
+          call error("User-specified onsite elements is not supported by DFTB+ API")
+       endif
+       
+    elseif(.not. tSccCalc)then
+       nIneqOrb = orb%nOrb
+       nMixElements = 0
+    end if
+
+  end subroutine updateEquivalencyRelations
+
+
+  !> Initialise partial charges
+  subroutine initialiseCharges(species0, speciesName, orb, nElectrons, iEqOrbitals, nIneqOrb, nMixElements, &
+       q0, qShell0, qInput, qInpRed, qOutput, qOutRed, qDiffRed, initialSpins, initialCharges)    
+
+    use dftbp_initprogram, only: tMulliken, tSccCalc, tSpin, tReadChrg, tCustomOccAtoms, tMixBlockCharges, tRangeSep, &
+                                 referenceN0, nSpin, nAtom  !, tImHam
+    use dftbp_sccinit,     only: initQFromShellChrg, initQFromAtomChrg
+    use dftbp_orbitalequiv,only: OrbitalEquiv_reduce
+    use dftbp_spin,        only: qm2ud,ud2qm
+    use dftbp_orbitals,    only: TOrbitals
+    use dftbp_accuracy,    only: mc
+    use dftbp_message,     only: error
+    
+    !> Type of the atoms (nAtom)
+    integer,  allocatable, intent(in) :: species0(:)
+    character(mc),  allocatable, intent(in) :: speciesName(:)
+    !> Data type for atomic orbitals
+    type(TOrbitals),       intent(in) :: orb
+    !> Number of electrons
+    real(dp),              intent(in) :: nElectrons(:)
+    !> Orbital equivalence relations
+    integer,  allocatable, intent(in) :: iEqOrbitals(:,:,:)
+    !> nr. of inequivalent orbitals
+    integer,               intent(in) :: nIneqOrb
+    !> nr. of elements to go through the mixer - may contain reduced orbitals and also orbital blocks                
+    !> (if tDFTBU or onsite corrections)                                           
+    integer, intent(in) :: nMixElements
+
+    !> Initial spins
+    real(dp), allocatable, intent(in), optional :: initialSpins(:,:)
+    !> Set of atom-resolved atomic charges 
+    real(dp), allocatable, intent(in), optional :: initialCharges(:)
+ 
+    !> reference neutral atomic occupations
+    real(dp), allocatable, intent(inout) :: q0(:, :, :)
+    !> shell resolved neutral reference
+    real(dp), allocatable, intent(inout) :: qShell0(:,:)
+    !> input charges (for potentials)
+    real(dp), allocatable, intent(inout) :: qInput(:, :, :)
+    !> output charges
+    real(dp), allocatable, intent(inout) :: qOutput(:, :, :)
+    !> input charges packed into unique equivalence elements
+    real(dp), allocatable, intent(inout) :: qInpRed(:)
+    !> output charges packed into unique equivalence elements
+    real(dp), allocatable, intent(inout) :: qOutRed(:)
+    !> charge differences packed into unique equivalence elements
+    real(dp), allocatable, intent(inout) :: qDiffRed(:)
+
+    !> TODO(Alex) Not global in initprogram. Currently hard-coded
+    !> Is check-sum for charges read externally to be used?
+    logical  :: tSkipChrgChecksum = .false.
```

Comment:
Pass this in

Reply:
Again, we should try to refactor the program in `initprogram.F90` and call it instead of duplicating things.